### PR TITLE
online covariance improvements

### PR DESCRIPTION
**Pull request by mchernov-intel**
# Description
Please include a summary of the change. For large or complex changes please include enough information to introduce your change and explain motivation for it.

Changes proposed in this pull request:
-
-
-

### CODE REVIEW

**Comment by mchernov-intel** [comment]
make test

**Comment by mchernov-intel** [comment]
make test

**Comment by mchernov-intel** [comment]
restart

**Comment by mchernov-intel** [comment]
restart

**Comment by mchernov-intel** [comment]
@michael-smirnov @PetrovKP please, review

File: service/kernel/oneapi/cl_kernels/vector_sum.cl [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+*/
+
+#ifndef __SUM_REDUCER_CL__
+#define __SUM_REDUCER_CL__
+
+#define DECLARE_SOURCE(name, src) static const char* name = #src;
+
+DECLARE_SOURCE(vector_sum,
+
+__kernel void vector_sum(__global const algorithmFPType *A,
```
Comment (PetrovKP) [inline service/kernel/oneapi/cl_kernels/vector_sum.cl]: Why write your crutches? Use axpy

File: service/kernel/oneapi/vector_sum.h [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+{
+namespace internal
+{
+namespace math
+{
+
+class VectorSum
+{
+public:
+    static void vectorSum(const UniversalBuffer& A,
```
Comment (PetrovKP) [inline service/kernel/oneapi/vector_sum.h]: variable A, B, etc do not match the style code

File: algorithms/kernel/covariance/oneapi/cl_kernels/covariance_kernels.cl [hunk 3]
```diff
@@ -48,31 +48,17 @@ __kernel void mergeCrossProduct(uint nFeatures,
...
-            algorithmFPType invPartialNObs = (algorithmFPType)(1.0) / partialNObservations;
-            algorithmFPType invNObs = (algorithmFPType)(1.0) / nObservations;
-            algorithmFPType invNewNObs = (algorithmFPType)(1.0) / (nObservations + partialNObservations);
-
-            crossProduct[i * nFeatures + j] += partialCrossProduct[i * nFeatures + j];
-            crossProduct[i * nFeatures + j] += partialSums[i] * partialSums[j] * invPartialNObs;
-            crossProduct[i * nFeatures + j] += sums[i] * sums[j] * invNObs;
-            crossProduct[i * nFeatures + j] -= (partialSums[i] + sums[i]) * (partialSums[j] + sums[j]) * invNewNObs;
-        }
+        algorithmFPType invPartialNObs = (algorithmFPType)(1.0) / partialNObservations;
```
Comment (PetrovKP) [inline algorithms/kernel/covariance/oneapi/cl_kernels/covariance_kernels.cl]: Use const

File: service/kernel/oneapi/build_program.h [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+#include "oneapi/internal/execution_context.h"
+
+namespace daal
+{
+namespace oneapi
+{
+namespace internal
+{
+
+static void buildProgram(ClKernelFactoryIface& kernelFactory,
```
Comment (michael-smirnov) [inline service/kernel/oneapi/build_program.h]: Why you declared this function `static` here?

Comment (michael-smirnov) [inline service/kernel/oneapi/build_program.h]: If you want us to use this function to build all the kernels inside the library, please, do such changes not only for `SumReducer`

File: service/kernel/oneapi/build_program.h [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    build_options.add("-cl-std=CL1.2 -D LOCAL_BUFFER_SIZE=256");
+
+    cachekey.add(fptype_name);
+    kernelFactory.build(ExecutionTargetIds::device,
+                        cachekey.c_str(),
+                        program,
+                        build_options.c_str());
+}
+
+} // namespackkjje internal
```
Comment (michael-smirnov) [inline service/kernel/oneapi/build_program.h]: `namespackkjje` 😄

File: service/kernel/oneapi/vector_sum.h [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+#ifndef __VECTOR_SUM_H__
+#define __VECTOR_SUM_H__
+
+#include "math_service_types.h"
+#include "services/buffer.h"
+#include "cl_kernels/vector_sum.cl"
+#include "service_defines_oneapi.h"
+#include "oneapi/internal/types_utils.h"
+#include "oneapi/internal/execution_context.h"
```
Comment (michael-smirnov) [inline service/kernel/oneapi/vector_sum.h]: Seems like you don`t need some of these includes here

File: service/kernel/oneapi/cl_kernels/vector_sum.cl (hunk 1 above)
Comment (mchernov-intel) [reply service/kernel/oneapi/cl_kernels/vector_sum.cl]: done

File: service/kernel/oneapi/build_program.h [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+{
+
+static void buildProgram(ClKernelFactoryIface& kernelFactory,
+                  const TypeId& vectorTypeId,
+                  services::String cachekey,
+                  const char *program)
+{
+    services::String fptype_name = getKeyFPType(vectorTypeId);
+    auto build_options = fptype_name;
+    build_options.add("-cl-std=CL1.2 -D LOCAL_BUFFER_SIZE=256");
```
[the remark this answers is not in the document]
Comment (mchernov-intel) [reply service/kernel/oneapi/build_program.h]: done

**Comment by mchernov-intel** [comment]
make test

**Comment by mchernov-intel** [comment]
restart

**Comment by mchernov-intel** [comment]
rerun

**Comment by mchernov-intel** [comment]
restart

**Comment by mchernov-intel** [comment]
restart

File: include/oneapi/internal/math/blas_executor.h [hunk 8]
```diff
@@ -215,12 +215,73 @@ class SyrkExecutor
...
+        {
+            auto x_buffer_t = x_buffer.template get<T>();
+            auto y_buffer_t = y_buffer.template get<T>();
+
+            services::Status statusAxpy;
+#ifdef ONEAPI_DAAL_USE_MKL_GPU_FUNC
+            MKLAxpy<T> functor(queue);
+            statusAxpy = functor(n, (T)a, x_buffer_t, incx, y_buffer_t, incy);
+#else
+            // TODO
```
Comment (PetrovKP) [inline include/oneapi/internal/math/blas_executor.h]: Need to do. In the nightly build you will not be work

File: include/oneapi/internal/math/blas_executor.h [hunk 9]
```diff
@@ -215,12 +215,73 @@ class SyrkExecutor
...
+            // ReferenceAxpy<T> functor;
+            // statusAxpy =
+            //     functor(transInv, trans, k, k, n, (T)alpha, a_buffer_t, lda, offsetA, a_buffer_t, lda, offsetA, (T)beta, c_buffer_t, ldc, offsetC);
+#endif
+
```
Comment (PetrovKP) [inline include/oneapi/internal/math/blas_executor.h]: Extra line

File: include/oneapi/internal/math/blas_executor.h (hunk 8 above)
Comment (mchernov-intel) [reply include/oneapi/internal/math/blas_executor.h]: done.

File: include/oneapi/internal/math/blas_executor.h (hunk 9 above)
Comment (mchernov-intel) [reply include/oneapi/internal/math/blas_executor.h]: I don't agree with you

**Comment by mchernov-intel** [comment]
make test

File: include/oneapi/internal/math/mkl_blas.h [hunk 10]
```diff
@@ -145,6 +145,50 @@ struct MKLSyrk
...
+                   cl::sycl::buffer<double, 1> & y_buffer, const int incy)
+    {
+        fpk::blas::axpy(_queue, n, a, x_buffer, incx, y_buffer, incy);
+    }
+
+    template <>
+    void innerAxpy(const int n, const float a, cl::sycl::buffer<float, 1> & x_buffer, const int incx,
+                   cl::sycl::buffer<float, 1> & y_buffer, const int incy)
+    {
+        fpk::blas::axpy(_queue, n, a, x_buffer, incx, y_buffer, incy);
```
Comment (PetrovKP) [inline include/oneapi/internal/math/mkl_blas.h]: You duplicate code from 178 line. Why?

**Comment by mchernov-intel** [comment]
restart

**Comment by mchernov-intel** [comment]
restart

**Comment by mchernov-intel** [comment]
restart

**Comment by mchernov-intel** [comment]
restart

**Comment by mchernov-intel** [comment]
restart

File: service/kernel/oneapi/cl_kernels/kernel_blas.cl [hunk 11]
```diff
@@ -123,6 +123,18 @@ __kernel void blas_sgemm(const uint k,
...
-static const size_t clKernelLen = strlen(clKernelGemm);
+DECLARE_SOURCE(clKernelAxpy,
 
-#endif // __GEMM_KERNELS_CL__
+__kernel void blas_axpy(const algorithmFPType a,
+                        const __global algorithmFPType *x,
+                        const int incx,
+                        __global algorithmFPType *y,
+                        const int incy) {
+    const int i_x = (get_local_id(0)) * incx;
```
Comment (PetrovKP) [inline service/kernel/oneapi/cl_kernels/kernel_blas.cl]: Why `get_local_id` instead `get_global_id` ?

File: include/oneapi/internal/math/blas_executor.h [hunk 12]
```diff
@@ -210,12 +211,72 @@ class SyrkExecutor
...
+        {}
+
+        template <typename T>
+        void operator()(Typelist<T>)
+        {
+            auto x_buffer_t = x_buffer.template get<T>();
+            auto y_buffer_t = y_buffer.template get<T>();
+
+            services::Status statusAxpy;
+#ifdef ONEAPI_DAAL_USE_MKL_GPU_FUNC
```
Comment (PetrovKP) [inline include/oneapi/internal/math/blas_executor.h]: Need use new macros.

**Comment by mchernov-intel** [comment]
restart

**Comment by mchernov-intel** [comment]
restart

**Comment by mchernov-intel** [comment]
@michael-smirnov please, review

**Comment by mchernov-intel** [comment]
restart

**Comment by mchernov-intel** [comment]
@michael-smirnov please, review

**Comment by michael-smirnov** [comment]
make test

File: include/oneapi/internal/execution_context.h [hunk 13]
```diff
@@ -339,6 +339,9 @@ class ExecutionContextIface
     virtual void syrk(math::UpLo upper_lower, math::Transpose trans, size_t n, size_t k, double alpha, const UniversalBuffer & a_buffer, size_t lda,
                       size_t offsetA, double beta, UniversalBuffer & c_buffer, size_t ldc, size_t offsetC, services::Status * status = NULL) = 0;
 
+    virtual void axpy(const uint32_t n, const double a, const UniversalBuffer x_buffer, const int incx,
+                                 const UniversalBuffer y_buffer, const int incy, services::Status * status = NULL) = 0;
```
Comment (michael-smirnov) [inline include/oneapi/internal/execution_context.h]: indentation

File: include/oneapi/internal/math/blas_executor.h [hunk 14]
```diff
@@ -210,12 +211,72 @@ class SyrkExecutor
...
+        template <typename T>
+        void operator()(Typelist<T>)
+        {
+            auto x_buffer_t = x_buffer.template get<T>();
+            auto y_buffer_t = y_buffer.template get<T>();
+
+            services::Status statusAxpy;
+#ifdef ONEAPI_DAAL_NO_MKL_GPU_FUNC
+            ReferenceAxpy<T> functor;
+            statusAxpy =
```
Comment (michael-smirnov) [inline include/oneapi/internal/math/blas_executor.h]: Align style in lines 253 and 257

File: service/kernel/oneapi/cl_kernels/kernel_blas.cl [hunk 15]
```diff
@@ -123,6 +123,18 @@ __kernel void blas_sgemm(const uint k,
...
-static const size_t clKernelLen = strlen(clKernelGemm);
+DECLARE_SOURCE(clKernelAxpy,
 
-#endif // __GEMM_KERNELS_CL__
+__kernel void blas_axpy(const algorithmFPType a,
+                        const __global algorithmFPType *x,
+                        const int incx,
+                        __global algorithmFPType *y,
+                        const int incy) {
+    const int i_x = (get_global_id(0)) * incx;
```
Comment (michael-smirnov) [inline service/kernel/oneapi/cl_kernels/kernel_blas.cl]: You do not check that `i_x` and `i_y` does not go over `n`

**Comment by mchernov-intel** [comment]
make test

**Pull request merged**